BK-569 Add AddressFieldModal

jfplataroti opened the pull request:
## Description

Added address field modal `AddressFieldModal` to allow the user to add/edit an address.

It doesn't store or saves any data, ideally, this component is used to get a new address or edit one. So the submit handler should provide the new/updated address, delegating the saving data process. 

### Scenarios:

When it's used without the `address` prop it will:
1) allow the user to select an address from the Google autocomplete
2) allow the user to edit the details of the address.

When it's used with the `address` prop it will:
1) allow the user to edit the details of the address.

## Examples

For adding a new address
```tsx
      <AddressFieldModal establishmentId={1} />
```

For editing an address
```tsx
<AddressFieldModal establishmentId={1} address={someAddress} />
```

Sample of the modal without an address:
![modal-address1](https://user-images.githubusercontent.com/1827213/106044099-477e4b80-60ad-11eb-9df4-cddd08d1bb18.gif)



## Code changes:
- Added `DeliveryService` in the services package.
- Added `AddressDetails` to allow the user to modify the address. (it uses a library for forms that works pretty nice `formik`).
- Added `OutOfBoundariesError` to show the modal error.

jfplataroti commented on packages/web/components/addressDetails/addressDetails.tsx:
```diff
@@ -0,0 +1,83 @@
+  onSubmit,
+  onClearAddress,
+}) => {
+  const formik = useFormik({
+    initialValues: {
+      address,
+      unit,
+      businessName,
+    },
+    onSubmit: async (values) => {
+      if (!onSubmit) {
+        return;
+      }
+
+      try {
+        await onSubmit(values);
+      } catch (e) {}
+    },
+  });
```
use formik to handle the value changes and form submission.

djoyal replied:
Just out of curiosity, why use the `Formik` stuff instead of `Grommet`'s builtin stuff? It looks like it can handle most of the "form plumbing" automatically (https://v2.grommet.io/form)

jfplataroti replied:
Nice!! sorry I didn't see this one. If we can rely on grommet for that even better :) 

I'll change it to Grommet.

jfplataroti commented on packages/web/components/addressField/addressField.tsx:
```diff
@@ -1,44 +1,70 @@
+    setValue(e);
+    if (e && onAddressSelected) {
+      const address = await convertToAddress(e.value);
+      await onAddressSelected(address);
```
notifies parent component about the selected address as `Address` type.

jfplataroti commented on packages/web/components/addressField/convertAddress.ts:
```diff
@@ -0,0 +1,34 @@
+import {geocodeByPlaceId} from 'react-google-places-autocomplete';
+import {GoogleAddress} from './types';
+
+export async function convertToAddress(googleAddress: GoogleAddress): Promise<Address> {
```
Converts a google Place address into our `Address` model.

jfplataroti commented on packages/web/pages/profile/address/[id].tsx:
```diff
@@ -1,46 +1,14 @@
     <Box pad="small">
       <Heading level="1">Address</Heading>
-      <AddressForm labels={labels} />
+      <AddressDetails address="75 Dolores St." onSubmit={onSubmit} />
```
sorry test code, it can be ignored until we do those pages.

jfplataroti commented on packages/web/pages/profile/address/new.tsx:
```diff
@@ -1,22 +1,16 @@
   return (
-    <Box pad="small">
+    <Box pad="small" justify="start">
       <Heading level="1">New Address</Heading>
-      <AddressForm labels={labels} />
+      <AddressDetails />
```
same as above, test code.

djoyal commented on packages/services/src/DeliveryService.ts:
```diff
@@ -0,0 +1,44 @@
+  static async validateAddress(address: Address): Promise<any> {
+    const body: Partial<ValidateAddressRequest> = {
+      dropoffAddress: address,
+      companyEstablishmentId: 1,
```
This should probably be passed into the function as an argument (I realize we only have one location right now, but we'll be adding another pretty soon)

jfplataroti replied:
oh, good point. I'll update it.

jfplataroti replied:
done :)

jfplataroti commented:
@djoyal ready for review, I added the `establishmentId` as a parameter to the `DeliveryService.validate` fn and to the `AddressFieldModal` component props.

djoyal commented on packages/web/components/addressModal/addressError.tsx:
```diff
@@ -0,0 +1,29 @@
+          {t('add_delivery_address_error_title')}
+        </Heading>
+        <Paragraph>{t('add_delivery_address_error_content')}</Paragraph>
+        <img src={require('../../public/images/delivery-area-map.png')} />
```
Can we change this to a `<Image />` (from `Grommet`) for consistency please?

djoyal commented on packages/web/components/addressModal/addressError.tsx:
```diff
@@ -0,0 +1,29 @@
+          onClick={onClose}
+          primary
+          margin={{top: 'small'}}
+          label={t('okey')}></Button>
```
`okey` looks like a typo?

djoyal commented on packages/web/components/addressModal/addressFieldModal.tsx:
```diff
@@ -0,0 +1,62 @@
+  const {t} = useTranslation();
+  const [showOutOfBoundaries, setShowOutOfBoundaries] = useState(false);
+  const [newAddress, setNewAddress] = useState<Address>(address);
+  const title = !newAddress ? t('add_address') : t('address_details');
```
Nice!

djoyal commented on packages/web/components/addressModal/addressFieldModal.tsx:
```diff
@@ -0,0 +1,62 @@
+  };
+  const onCloseWarning = () => setShowOutOfBoundaries(false);
+  const onSubmitAddress = async (e) => {
+    console.log('submit form to be completed once is connected to a');
```
It looks like the end of the message is missing, can you finish it please?

djoyal commented on packages/web/components/loading.tsx:
```diff
@@ -0,0 +1,19 @@
+ * Placeholder for animation icon (or whatever we use).
+ */
+export const Loading: React.FunctionComponent = () => (
+  <Box
```
Nice!

djoyal commented on packages/web/public/static/locales/en/common.json:
```diff
@@ -1,5 +1,6 @@
 {
   "continue": "Continue",
+  "okey": "Okey",
```
I think this should be `okay` ?

jfplataroti replied:
yes heheh, thanks for catching them.

djoyal reviewed (commented):
This looks great overall! 

There are some small items that need cleanup, and I think we'll want to have a discussion about `Formik` vs `Grommet Forms` as well.

Nice work!

djoyal commented:
> @djoyal ready for review, I added the `establishmentId` as a parameter to the `DeliveryService.validate` fn and to the `AddressFieldModal` component props.

Thanks for making those changes! I completed the review and left some additional comments.

jfplataroti commented:
@djoyal thanks for finding the typos and realizing that grommet has a built-in Form handling!! :)  

I updated the PR with those changes:
- fixed typos in localization
- moved Form handler from formik to grommet.
- removed formik
- updated `Address` model to include address details (unit, businessName).
- updated `AddressDetail` props to use `Address`.

djoyal commented on packages/services/src/BKService.ts:
```diff
@@ -62,7 +62,10 @@ export class BKService {
     });
 
-    return response.json();
+    if (response.ok) {
```
Nice addition!

djoyal commented on packages/services/src/DeliveryService.ts:
```diff
@@ -0,0 +1,48 @@
+
+  static async validateAddress(
+    address: Address,
+    companyEstablishmentId: number
```
Thanks for moving this to an argument, it'll really help later on!

djoyal commented on packages/web/components/addressDetails/addressDetails.tsx:
```diff
@@ -0,0 +1,63 @@
+  };
+
+  const onChange = (nextValue) => setValue(nextValue);
+  console.log('v', value);
```
Can we take this out?

djoyal commented on packages/web/components/addressDetails/addressDetails.tsx:
```diff
@@ -0,0 +1,63 @@
+    ...address,
+  });
+
+  const onSubmitForm = async ({value}) => {
```
Thanks for switching this to the `Grommet Forms`! One less dependency to worry about :)

djoyal reviewed (approved):
Awesome, thanks for incorporating all the feedback @jfplataroti ! The PR looks really great!

There's one small nitpick, but nothing to worry about :)

jfplataroti commented:
Thanks for the feedback! 
I just removed the console.log and changed the `img` tag to the grommet one `Image`.
